### PR TITLE
Stack Editor: Logs view, editor improvements and stack create flow

### PR DESCRIPTION
**Pull request by gokmen**

### CODE REVIEW

**Comment by gokmen** [comment]
fixes: #10184

**Comment by gokmen** [comment]
fixes: #10186

File: client/app/lib/providers/computecontroller.coffee [hunk 1]
```diff
@@ -37,7 +37,7 @@ require './config'
 module.exports = class ComputeController extends KDController
 
 
-  @providers = globals.config.providers._getSupportedProviders()
+  PROVIDERS  = globals.config.providers._getSupportedProviders()
```
Comment (usirin) [inline client/app/lib/providers/computecontroller.coffee]: [minor] maybe extracting this to a util?

File: client/new-stack-editor/lib/controllers/base.coffee [hunk 2]
```diff
@@ -0,0 +1,10 @@
+kd = require 'kd'
+
+
+module.exports = class BaseController extends kd.Object
+
+  constructor: (options = {}, data) ->
+
+    super options, data
+
+    @editor = @getOption 'editor'
```
Comment (usirin) [inline client/new-stack-editor/lib/controllers/base.coffee]: Right now this only creates just another indirection with any visible gain (`@getOption 'editor'` is always available without this base class), but i am guessing you are thinking about extending this one's functionality.

File: client/new-stack-editor/lib/index.coffee [hunk 3]
```diff
@@ -36,27 +43,25 @@ module.exports = class StackEditorAppController extends AppController
     @stackEditor.on 'InitializeRequested', @bound 'initializeStack'
 
 
-  openEditor: (templateId) ->
+  openEditor: (templateId, reset = no) ->
```
Comment (usirin) [inline client/new-stack-editor/lib/index.coffee]: Can we change the 2nd argument to be an object, that way, we will have to explicitly write `{ reset: no }`. IMO it improves its readability, and removes the need for checking the implementation to understand what `no` or `yes` is used for at 2nd arg.

File: client/new-stack-editor/lib/index.coffee [hunk 4]
```diff
@@ -36,27 +43,25 @@ module.exports = class StackEditorAppController extends AppController
...
+    markAsLoaded()
 
 
-  reloadEditor: (templateId, skipDataUpdate) ->
+  reloadEditor: (templateId) ->
 
-    console.trace()
-    log '::reloadEditor', templateId, skipDataUpdate
+    delete @templates[templateId]
+    @openEditor templateId, reset = yes
```
Comment (usirin) [inline client/new-stack-editor/lib/index.coffee]: aha, you already felt the same but applied another solution 🤔 
i liked it, but i am still in favor of using an object, to be less error prone.

File: client/new-stack-editor/lib/views/editor.coffee [hunk 5]
```diff
@@ -18,15 +21,40 @@ module.exports = class Editor extends BaseView
...
+
   focus: -> @ready =>
     @aceView.ace.focus()
 
 
+  setReadOnly: (state) -> @ready =>
+
+    @aceView.ace.setReadOnly state, no
+    state = if state then 'none' else 'block'
+    @aceView.ace.editor.renderer.$cursorLayer.element.style.display = state
```
Comment (usirin) [inline client/new-stack-editor/lib/views/editor.coffee]: (titredim...) :D

File: client/new-stack-editor/lib/views/wizard/providerselectionview.coffee [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    @createProviders()
+
+
+  createProviders: ->
+
+    { providers } = globals.config
+
+    supportedProviders = (Object.keys providers).filter (p) ->
+      return p  if providers[p].supported
```
Comment (usirin) [inline client/new-stack-editor/lib/views/wizard/providerselectionview.coffee]:
```coffee
supportedProviders = (Object.keys providers).filter (p) -> providers[p].supported
```

actually with lodash there is an easier and -IMO- more readable way ([link to doc](https://lodash.com/docs/4.17.3#pickBy)):

```coffee
_.pickBy providers, (provider) -> provider.supported
```

All three of them do the same thing.

Comment (usirin) [inline client/new-stack-editor/lib/views/wizard/providerselectionview.coffee]: btw, i don't have any preferences, i just wanted to share my opinions. I hate doing `Object.keys ....` everytime.

File: client/new-stack-editor/lib/views/wizard/providerselectionview.coffee [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        click      : =>
+          return  if extraClass is 'coming-soon'
+
+          Tracker.track Tracker["STACKS_WIZARD_SELECTED_#{provider.toUpperCase()}"]
+
+          providerView.setClass 'selected'
+          @selected?.unsetClass 'selected'
+
+          @selected = if @selected is providerView then null else providerView
+          @emit 'SelectedProviderChanged', @selected
```
Comment (usirin) [inline client/new-stack-editor/lib/views/wizard/providerselectionview.coffee]: Recently i have started to extract out this kind of hardcoded values into a `constants.coffee` right next to the files that uses them. It's especially useful when one wants to change the value of the event, change in one place, and don't think about where it was being used.

File: client/new-stack-editor/lib/views/wizard/stackwizard.coffee [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+
+  pistachio: ->
+
+    '''
+      {{> @providerSelectionView}}
+      <footer>
+        {{> @createButton}}
+        {{> @cancelButton}}
+      </footer>
```
Comment (usirin) [inline client/new-stack-editor/lib/views/wizard/stackwizard.coffee]: Unnecessary indentation.

File: client/new-stack-editor/lib/views/wizard/stackwizardmodal.coffee [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+      @on 'KDObjectWillBeDestroyed', -> kd.singletons.router.back()
+
+    @addSubView wizard
+
+
+  createStackTemplate: (selectedProvider) ->
+
+    { router } = kd.singletons
+
+    unless selectedProvider
```
Comment (usirin) [inline client/new-stack-editor/lib/views/wizard/stackwizardmodal.coffee]: Why and how do we need this case?

**Review by usirin — APPROVED** [review]
My comments are mostly minors/nitpicks. We can tackle those issues later. Merging.

File: client/new-stack-editor/lib/index.coffee (hunk 3 above)
Comment (gokmen) [inline client/new-stack-editor/lib/index.coffee]: I usually not prefer that if there is only one argument, if there will be more than I can update.

File: client/new-stack-editor/lib/views/editor.coffee (hunk 5 above)
Comment (gokmen) [inline client/new-stack-editor/lib/views/editor.coffee]: this is the only way to hide cursor on Ace, it doesn't look good I agree. there are terrible solutions for that one believe me :) https://github.com/ajaxorg/ace/issues/266

File: client/new-stack-editor/lib/views/wizard/providerselectionview.coffee (hunk 6 above)
Comment (gokmen) [inline client/new-stack-editor/lib/views/wizard/providerselectionview.coffee]: I agree but I usually prefer to use the way everybody can understand without a knowledge. TIL `pickBy` we can use it further.

File: client/new-stack-editor/lib/views/wizard/stackwizardmodal.coffee (hunk 9 above)
Comment (gokmen) [inline client/new-stack-editor/lib/views/wizard/stackwizardmodal.coffee]: my bad, in the previous example there was a case for this one but it's not valid anymore.

File: client/app/lib/providers/computecontroller.coffee (hunk 1 above)
Comment (gokmen) [inline client/app/lib/providers/computecontroller.coffee]: most of the time where I need supported providers list I also need providers as well so I kept it in the same object for consistency.

File: client/new-stack-editor/lib/controllers/base.coffee (hunk 2 above)
Comment (gokmen) [inline client/new-stack-editor/lib/controllers/base.coffee]: in this file I may re-set the option and delete editor from there after I've assigned it to local variable. so, there are some plans on `@editor` variable for the future.

**Pull request merged**